### PR TITLE
Handle CORS at ASGI level

### PR DESCRIPTION
**Pull request by patrys**
This is faster than doing it inside the Django view but, like the health check, it won't work with `manage.py runserver`.

⚠️ Django's `runserver` command does not support ASGI yet, so if you want CORS to work on localhost, you'll need to run the ASGI server instead:

```
python -m saleor
```

or

```
uvicorn saleor.asgi:application --reload
```

Please note that this change will be needed anyway once we start using `asyncio` more.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
I don't know what to do with tests as they currently don't use ASGI either.

**Comment by maarcingebala** [comment]
- We should add changelog entry.
- Shouldn't this be considered as a breaking change for OSS users? Is there a change someone is using a different server to run Saleor that doesn't support ASGI?
- As for the tests, I see Django has an [async client](https://docs.djangoproject.com/en/4.1/topics/testing/tools/#django.test.AsyncClient) can't we use it? Or is it a more general problem with Pytest not supporting async?

**Comment by patrys** [comment]
Saleor can’t support both WSGI and ASGI if we want to take advantage of asyncio. I’ll see if I can get Pytest to play nice with ASGI, otherwise I can just test the ASGI middleware directly.

**Comment by patrys** [comment]
@maarcingebala Switched tests to test the ASGI middleware explicitly.

**Review by fowczarek — APPROVED** [review]

**Comment by fowczarek** [comment]
@patrys Should we change the command in [platform](https://github.com/saleor/saleor-platform/blob/main/docker-compose.yml#L27-L28)?

**Comment by patrys** [comment]
@fowczarek I will if we decide we want to merge this.

**Comment by patrys** [comment]
Here's a PR to use ASGI in `saleor-platform`: https://github.com/saleor/saleor-platform/pull/225

File: CHANGELOG.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ All notable, unreleased changes to this project will be documented in this file.
 - Remove `CUSTOMER_UPDATED` webhook trigger from address mutations - #11395 by @jakubkuc
 - Drop `Django.Auth` - #11305 by @fowczarek
 - Propagate voucher discount between checkout lines when charge_taxes is disabled - #11632 by @maarcingebala
+- CORS is now handled in the ASGI layer - #11415 by @patrys
```
Comment (maarcingebala) [inline CHANGELOG.md:33]: Need to be moved to 3.11 section.

**Review by maarcingebala — APPROVED** [review]

**Comment by patrys** [comment]
Rebased on the current main branch

**Comment by maarcingebala** [comment]
@patrys It needs rebase again 🙏

**Comment by omars-code** [comment]
This solves the OPTIONS but not the POST

**Comment by omars-code** [comment]
just an FYI there is this fork, but the package seems abandoned 
https://github.com/omars-code/asgi-cors

**Comment by patrys** [comment]
@omars-code Good catch, see https://github.com/saleor/saleor/pull/12000

**Pull request merged**